contracts/periphery/contracts/s2s/MessageEndpoint.sol

```diff
@@ -0,0 +1,246 @@
+        uint32 tgtSpecVersion,
+        address callReceiver,
+        bytes calldata callPayload,
+        uint256 gasLimit
+    ) internal returns (uint256) {
+        bytes memory input = abi.encodeWithSelector(
+            this.execute.selector,
+            callReceiver,
+            callPayload
+        );
+
+        return _remoteTransact(tgtSpecVersion, input, gasLimit);
+    }
+
+    function _remoteTransact(
+        uint32 tgtSpecVersion,
+        bytes memory input,
+        uint256 gasLimit
+    ) internal returns (uint256) {
+        PalletEthereum.MessageTransactCall memory call = PalletEthereum
+            .MessageTransactCall(
+                // the call index of message_transact
+                remoteMessageTransactCallIndex,
+                // the evm transaction to transact
+                PalletEthereum.buildTransactionV2ForMessageTransact(
+                    gasLimit,
+                    remoteEndpoint,
+                    remoteSmartChainId,
+                    input
+                )
+            );
+        bytes memory callEncoded = PalletEthereum.encodeMessageTransactCall(
+            call
+        );
+        uint64 weight = uint64(gasLimit * remoteWeightPerGas);
+
+        return _remoteDispatch(tgtSpecVersion, callEncoded, weight);
+    }
+
+    function _remoteDispatch(
+        uint32 tgtSpecVersion,
+        bytes memory tgtCallEncoded,
+        uint64 tgtCallWeight
+    ) internal returns (uint256) {
+        // Build the encoded message to be sent
+        bytes memory message = SmartChainXLib.buildMessage(
+            tgtSpecVersion,
+            tgtCallWeight,
+            tgtCallEncoded
+        );
+
+        // Send the message
+        SmartChainXLib.sendMessage(
+            dispatchAddress,
+            sendMessageCallIndex,
+            outboundLaneId,
+            msg.value,
+            message
+        );
+
+        // Get nonce from storage
+        uint64 nonce = SmartChainXLib.latestNonce(
+            storageAddress,
+            storageKeyForLatestNonce,
+            outboundLaneId
+        );
+
+        return encodeMessageId(outboundLaneId, nonce);
+    }
+
+    ///////////////////////////////
+    // Inbound
+    ///////////////////////////////
+    modifier onlyMessageSender() {
+        require(
+            derivedMessageSender == msg.sender,
+            "MessageHandle: Invalid sender"
+        );
+        _;
+    }
+
+    function execute(address callReceiver, bytes calldata callPayload)
+        external
+        onlyMessageSender
+    {
+        if (_canBeExecuted(callReceiver, callPayload)) {
+            (bool success, ) = callReceiver.call(callPayload);
+            require(success, "MessageHandle: Call execution failed");
+        } else {
+            revert("MessageHandle: Unapproved call");
+        }
+        
+    }
+
+    // Check if the call can be executed
+    function _canBeExecuted(address callReceiver, bytes calldata callPayload)
```

Comment:
the `approved` term is often used in smart contract area. so, I replaced `_approved` with `_canBeExecuted `. @xiaoch05